drivers/clock_control/clock_control_esp32.c

```diff
@@ -0,0 +1,318 @@
+{
+	ARG_UNUSED(dev);
+	u32_t bank = GET_REG_BANK(sys);
+	u32_t offset =  GET_REG_OFFSET(sys);
+
+	__ASSERT_NO_MSG(bank >= CLOCK_REGS_BANK_COUNT);
+
+	esp32_set_mask32(BIT(offset), clock_control_regs[bank].clk);
+	esp32_clear_mask32(BIT(offset), clock_control_regs[bank].rst);
+	return 0;
+}
+
+static int clock_control_esp32_off(struct device *dev,
+				   clock_control_subsys_t sys)
+{
+	ARG_UNUSED(dev);
+	u32_t bank = GET_REG_BANK(sys);
+	u32_t offset =  GET_REG_OFFSET(sys);
+
+	__ASSERT_NO_MSG(bank >= CLOCK_REGS_BANK_COUNT);
+
+	esp32_clear_mask32(BIT(offset), clock_control_regs[bank].clk);
+	esp32_set_mask32(BIT(offset), clock_control_regs[bank].rst);
+	return 0;
+}
+
+static enum clock_control_status clock_control_esp32_get_status(struct device *dev,
+								clock_control_subsys_t sys)
+{
+	ARG_UNUSED(dev);
+	u32_t bank = GET_REG_BANK(sys);
+	u32_t offset =  GET_REG_OFFSET(sys);
+
+	if (DPORT_GET_PERI_REG_MASK(clock_control_regs[bank].clk, BIT(offset))) {
+		return CLOCK_CONTROL_STATUS_ON;
+	}
+	return CLOCK_CONTROL_STATUS_OFF;
+}
+
+static int clock_control_esp32_get_rate(struct device *dev,
+					clock_control_subsys_t sub_system,
+					u32_t *rate)
+{
+	ARG_UNUSED(sub_system);
+
+	u32_t xtal_freq_sel = DEV_CFG(dev)->xtal_freq_sel;
+	u32_t soc_clk_sel = REG_GET_FIELD(RTC_CNTL_CLK_CONF_REG, RTC_CNTL_SOC_CLK_SEL);
+
+	switch (soc_clk_sel) {
+	case RTC_CNTL_SOC_CLK_SEL_XTL:
+		*rate = xtal_freq[xtal_freq_sel];
+		return 0;
+	case RTC_CNTL_SOC_CLK_SEL_PLL:
+		*rate = MHZ(80);
+		return 0;
+	default:
+		*rate = 0;
+		return -ENOTSUP;
+	}
+}
+
+static int clock_control_esp32_init(struct device *dev)
+{
+	struct esp32_clock_config *cfg = DEV_CFG(dev);
+
+	/* Wait for UART first before changing freq to avoid garbage on console */
+	esp32_rom_uart_tx_wait_idle(0);
+
+	switch (cfg->clk_src_sel) {
+	case ESP32_CLK_SRC_XTAL:
+		REG_SET_FIELD(APB_CTRL_SYSCLK_CONF_REG, APB_CTRL_PRE_DIV_CNT, cfg->xtal_div);
+		/* adjust ref_tick */
+		REG_WRITE(APB_CTRL_XTAL_TICK_CONF_REG, xtal_freq[cfg->xtal_freq_sel] - 1);
+		/* switch clock source */
+		REG_SET_FIELD(RTC_CNTL_CLK_CONF_REG, RTC_CNTL_SOC_CLK_SEL, RTC_CNTL_SOC_CLK_SEL_XTL);
+		break;
+	case ESP32_CLK_SRC_PLL:
+		cpuclk_pll_configure(cfg->xtal_freq_sel, cfg->cpu_freq);
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	/* Re-calculate the CCOUNT register value to make time calculation correct.
+	 * This should be updated on each frequency change
+	 * New CCOUNT = Current CCOUNT * (new freq / old freq)
+	 */
+	XTHAL_SET_CCOUNT((u64_t)XTHAL_GET_CCOUNT() * cfg->cpu_freq / xtal_freq[cfg->xtal_freq_sel]);
+	return 0;
+}
+
+static const struct clock_control_driver_api clock_control_esp32_api = {
+	.on = clock_control_esp32_on,
+	.off = clock_control_esp32_off,
+	.get_rate = clock_control_esp32_get_rate,
+	.get_status = clock_control_esp32_get_status,
+};
+
+static const struct esp32_clock_config esp32_clock_config0 = {
+	.clk_src_sel = DT_PROP(DT_INST(0, cadence_tensilica_xtensa_lx6), clock_source),
```

Comment:
To make this optional in the binding description, use something like this:
```
#define CPU_NODE DT_INST(0, cadence_tensilica_xtensa_lx6)

.clk_src_sel = COND_CODE_1(DT_NODE_HAS_PROP(CPU_NODE, clock_source),
                           (DT_PROP(CPU_NODE, clock_source)),
			   ESP32_CLK_SRC_PLL),
```

though I haven't tested that so it may be slightly wrong.

Reply:
this should be

```
COND_CODE_1(DT_NODE_HAS_PROP(CPU_NODE, clock_source),
            (DT_PROP(CPU_NODE, clock_source)),
            (ESP32_CLK_SRC_PLL))
```

Note the extra parentheses around ESP32_CLK_SRC_PLL. This is required by COND_CODE_1.

We didn't have time in zephyr 2.3 to add macros for "get me this property value if it exists, or this default value otherwise". We should add them for 2.4.

We should do the same for "get me this cell value if it exists, or this default otherwise", and reasonable extensions thereof, like "get me this flags value **or zero**".

I'll add that sometime during the next development cycle. https://github.com/zephyrproject-rtos/zephyr/issues/25827

Reply:
Thanks for Guidance.

I decided in the end to add `BUILD_ASSERT` for this property, it will be better to give compilation error in this, because No one needs a system with incorrect clock settings.

If i want to make this work, I had to add multiple additional checks which IMHO are not necessary.